feat: key value store backed by filesystem (#20932)
microsoft/playwright: docs/src/test-api/class-teststore.md

# class: TestStore
* since: v1.32
* langs: js

> Playwright Test provides a global `store` object that can be used read/write values to the filesystem. Each value is stored in its own file with the key used as a path relative to `TestConfig.storeDir` (it is `./playwright` by default).

Reviewer: ```suggestion
Playwright Test provides a global `store` object that can be used to read/write values on the filesystem. Each value is stored in its own file inside './playwright' directory, configurable with [`property: TestConfig.storeDir`].
```

```ts
import { test, store } from '@playwright/test';

test('get user name', async ({ page, context }) => {
  await page.goto('/');
  // Return mock user info from the store.
  await page.route('**/info/user', route => route.fulfill({ path: store.path('mocks/user.json')}))
  await page.getByText('My Profile');
  // Check that the name matches mock data.
  await expect(page.getByLabel('Name')).toHaveText('John');
});
```

> 

Reviewer: Suggestion: `method: TestStore.root` to get the root?

Author: Let's wait for now?

Author: Added.

## async method: TestStore.get
* since: v1.32
- returns: <[any]>

> Get named item from the store. Returns undefined if there is no value with given path.

Reviewer: We should figure out:
- whether we support binary values like images;
- whether we treat json specially and JSON.parse it;
- whether we detect mime type from extension or pass it explicitly.

Author: > whether we support binary values like images;
We should, I'll address this separately.

> whether we treat json specially and JSON.parse it;
I'd say yes, as this is expected to be the most common use case.

> whether we detect mime type from extension or pass it explicitly.
If the keys include extension like in this implementation to we should infer the mime type from it I believe. In any case I'll tackle this in a separate PR.

### param: TestStore.get.path
* since: v1.32
- `path` <[string]>

Item path.

## method: TestStore.path
* since: v1.32
- returns: <[string]>

Returns absolute path of the corresponding store entry on the file system.

### param: TestStore.path.path
* since: v1.32
- `path` <[string]>

Path of the item in the store.

## async method: TestStore.set
* since: v1.32

Set value to the store.

### param: TestStore.set.path
* since: v1.32
- `path` <[string]>

Item path.

### param: TestStore.set.value
* since: v1.32
- `value` <[any]>

> Item value. The value must be serializable to JSON. Passing `undefined` deletes the entry with given path.

Reviewer: I would also do `delete` right away.
